proposals/NNNN-inheritance-of-actor-isolation.md

```diff
@@ -0,0 +1,653 @@
+  implicit `self` of an actor method) and that parameter is strongly
+  captured by the closure.
+
+The third clause is modified by this proposal to say that isolation
+is also inherited if a non-optional binding of an isolated parameter
+is captured by the closure (see below).
+
+### Generalized isolation checking
+
+When calling a function with an `isolated` parameter, the function
+shares the same isolation as the current context if:
+
+- the current context is non-isolated, the parameter type is optional,
+  and the argument expression is `nil` or a reference to `Optional.none`;
+
+- the current context has an `isolated` parameter (including the
+  implicitly-`isolated` `self` parameter of an actor function) and
+  the argument expression is a reference to that parameter or a
+  non-optional derivation of it (see below); or
+
+- the current context is isolated to a global actor type `T` and the
+  argument expression is `T.shared`, where `shared` is `GlobalActor`'s
+  protocol requirement or the concrete declaration which provides it
+  in `T`'s conformance to `GlobalActor`.
+
+An expression is a non-optional derivation of an isolated parameter
+`param` if it is:
+- `param?` (the optional-chaining operator);
+- `param!` (the force-unwrapping operator); or
+- a reference to a *non-optional binding* of `param`, i.e. a `let`
+  constant initialized by a successful pattern-match which removes
+  the optionality from `param`, such as `ref` in `if let ref = param`.
+
+When analyzing an argument expression in all cases above, certain
+non-instrumental differences in expression syntax and behavior must
+be ignored:
+- parentheses;
+- the effect-marking operators `try`, `try?`, `try!`, and `await`;[^5]
+- the type coercion operator `as` (in the cases where it doesn't
+  perform a dynamic bridging conversion); and
+- implicit type conversions such as promotion to `Optional` type.
+
+[^5]: The restrictions on the underlying expression should make it
+pointless to use these operators, but they must be ignored anyway.
+
+Note that the special `#isolation` default argument form should
+always be replaced by something matching the rule above, so calls
+using this default argument for an isolated parameter will always be
+to a context that shares isolation.
+
+Calls to `@inheritsIsolation` functions and initializers, and uses of
+`@inheritsIsolation` variables and subscripts, always share the isolation
+of the current context.
+
+For example:
+
+```swift
+/// This class type is not Sendable.
+class Counter {
+  var count = 0
+}
+
+extension Counter {
+  /// Since this is an async function, if it were just declared
+  /// non-isolated, calling it from an isolated context would be
+  /// forbidden because it requires sharing a non-Sendable value
+  /// between concurrency domains.  Inheriting isolation makes it
+  /// okay.  This is a contrived example chosen for its simplicity.
+  func incrementAndSleep(isolation: (any Actor)?) async {
```

Comment:
I think this is missing an isolated? Otherwise this is nonisolated actually

```suggestion
  func incrementAndSleep(isolation: isolated (any Actor)?) async {
```